### ISSUE TITLE
sweeper: `SignOutputRaw` called without UTXO pk script

### ISSUE THREAD

**Issue by guggero** [lightninglabs/loop#457]
Found in combination with remote signing, see https://github.com/lightningnetwork/lnd/issues/6227.

In https://github.com/lightninglabs/loop/blob/master/sweep/sweeper.go#L64 we create the sign descriptor for calling `SignOutputRaw` but we don't include the UTXOs pk script, only the value.
This works fine in the normal (full private keys available) wallet but presents a problem in a remote-signing setup where the signing information needs to be packaged in a PSBT to be sent over to the signing instance. An empty pk script field in the `WitnessUTXO` field of a PSBT is invalid (apparently).

I checked the code and we have the HTLC pk script in all cases where we use the sweeper, so it should be relatively easy to pass it into the call.

**Comment by alexbosworth** [comment]
is the pk script always available in all versions of lnd?

**Comment by guggero** [comment]
Yes. The pk script is the hash of the HTLC script. So loop always has it, it just needs to send it over to `lnd` in the `SignOutputRaw` call as part of the field in the sign descriptor that describes the UTXO to sign.

**Comment by mrtestyboy781** [comment]
Thanks for making this issue, is there any particular hot fix I can apply to get this working?

**Comment by mrtestyboy781** [comment]
To follow up, I tried https://github.com/mrtestyboy781/loop/commit/15d231c787f58e80be6494895f5633232a950d2e but was getting 

```
2022-03-08 13:37:29.473 [WRN] BTWL: SignPsbt: Skipping input 0, derived public key 03a02df82e9cf979e077f897b57a864aa3040a7bc8fdc60b13aba424f431aa3f79 does not match bip32 derivation info public key 02f9a06202d8b812e39593ac06695b9552e23687c28badeb9627c4ea10e82c73a3
```

On the remote signer, so I guess im still not setting the pk script quite right.

**Comment by mrtestyboy781** [comment]
@guggero - based on the description of the issue, i halfway expected the above commit might do the trick, but that doesn't seem to be the case. I'm happy to test a fix and create an MR if you can give me a hint for whats going wrong.

**Comment by guggero** [comment]
Hmm, I think this might be another problem with the `SignOutputRaw` RPC. Only setting the public key in the `KeyDescriptor` should work, but apparently doesn't. Can you try also setting the key family and index? If that info is available on that level...

**Comment by mrtestyboy781** [comment]
@guggero did some more digging....

It seems if you pass the whole KeyDescriptor (with KeyLocator) [here](https://github.com/lightninglabs/loop/blob/master/sweep/sweeper.go#L68-L70) - the only thing that gets serialized to [signer_server](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/signrpc/signer_server.go#L241) is `keyDesc : raw_key_bytes:"..."`  and not the KeyLocator

This is a problem because later in [remoteSign](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/rpcwallet/rpcwallet.go#L571) the KeyDesc is set to `Key Desc: {KeyLocator:{Family:0 Index:0}` (KeyFamily should be 99 for loop) and the [bip32Derivation is then set wrong based on that](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/rpcwallet/rpcwallet.go#L607-L620) 

In trying to figure out why KeyLocator isn't passed onto `SignOutputRaw` 
I thought it might be because [SignerDescriptor serialization](https://github.com/lightningnetwork/lnd/blob/master/input/signdescriptor.go#L89-L96) tries to read `sd.KeyDesc.Family` instead of `sd.KeyDesc.KeyLocator.Family` and `sd.KeyDesc.Index` instead of `sd.KeyDesc.KeyLocator.Index` - but that didn't seem to help. 

Wondering if its a mismatch in proto buf between expecting `KeyLoc` vs `KeyLocator` - or something else simple. I really dont know the codebase at all, or go for that matter

Loop debug : https://github.com/mrtestyboy781/loop/commit/c53aa8688850166ce9c720d3c296742874ad6f74
Lnd debug : https://github.com/mrtestyboy781/lnd/commit/c9e8a0f78f70ca08d73b72bd0898291676742e5e

**Comment by guggero** [comment]
Great, thank you for digging into this further! I plan on taking a closer look myself tomorrow. So far, from what you found, it sounds to me like we're just not passing along some required information in the remote signing setup.

**Comment by guggero** [comment]
Okay, I think I figured it out. It should be enough to just call the `SignOutputRaw` RPC with the public key as Loop does. But in the case of the remote signing setup, some information isn't passed along from the watch-only node to the signer.
This needs a fix in `lnd` itself, which I should have up by tomorrow, I hope.
So Loop only needs to provide the pkScript as mentioned in this issue, no other changes should be necessary. But we'll make sure Loop does run with a remote signing backend for the next release (of both `lnd` and Loop).

**Comment by mrtestyboy781** [comment]
Great, thank you

**Comment by guggero** [comment]
Just a heads up that even though this was merged, running Loop against a remote signing `lnd` pair will only work if the watch-only node is on a version of `lnd` that includes https://github.com/lightningnetwork/lnd/pull/6341 (which will hopefully be the case with the soon-to-be-released `v0.14.3-beta`).
